### PR TITLE
Charlesmchen/share contacts from share extension

### PR DESCRIPTION
**Pull request by charlesmchen-signal**
This PR is now ready for review. 

PTAL @michaelkirk

### CODE REVIEW

File: SignalMessaging/attachments/ApproveContactShareViewController.swift [hunk 1]
```diff
@@ -277,7 +296,16 @@ public class ApproveContactShareViewController: OWSViewController, EditContactSh
 
     // TODO: Surface error with resolution to user if not.
```
Comment (charlesmchen-signal) [inline SignalMessaging/attachments/ApproveContactShareViewController.swift]: I believe this PR contains all of the SAE-specific work for contact sharing. Any remaining work is in classes used in contact sharing in main app and SAE.

File: SignalMessaging/attachments/SignalAttachment.swift [hunk 2]
```diff
@@ -142,6 +142,10 @@ public class SignalAttachment: NSObject {
     @objc
     public var isConvertibleToTextMessage = false
 
+    // This flag should be set for attachments that can be sent as contact shares.
+    @objc
+    public var isConvertibleToContactShare = false
```
Comment (charlesmchen-signal) [inline SignalMessaging/attachments/SignalAttachment.swift]: These probably belong in a SAE-specific category for this class.

File: SignalShareExtension/Info.plist [hunk 3]
```diff
@@ -60,11 +60,8 @@
                 SUBQUERY (
                 $extensionItem.attachments,
                 $attachment,
-                (
                 ANY $attachment.registeredTypeIdentifiers UTI-CONFORMS-TO "public.data"
                 || ANY $attachment.registeredTypeIdentifiers UTI-CONFORMS-TO "public.url"
-                )
-                AND NOT (ANY $attachment.registeredTypeIdentifiers UTI-CONFORMS-TO "public.vcard")
```
Comment (charlesmchen-signal) [inline SignalShareExtension/Info.plist]: @michaelkirk  Please scrutinize this carefully.

File: SignalShareExtension/ShareViewController.swift [hunk 4]
```diff
@@ -719,6 +720,15 @@ public class ShareViewController: UIViewController, ShareViewDelegate, SAEFailed
                 if ShareViewController.itemMatchesSpecificUtiType(itemProvider: itemProvider,
                                                                   utiType: kUTTypeVCard as String) {
                     customFileName = "Contact.vcf"
+
+                    if let contactShare = OWSContacts.contact(forVCardData: data) {
+                        isConvertibleToContactShare = true
+                    } else {
+                        Logger.error("\(strongSelf.logTag) could not parse vcard.")
+                        let writeError = ShareViewControllerError.assertionError(description: "Could not parse vcard data.")
+                        reject(writeError)
```
Comment (charlesmchen-signal) [inline SignalShareExtension/ShareViewController.swift]: The contact-conversion logic is pretty permissive. This should fail only for corrupt vcards or vcards that contain no contacts.  If it contains even a single contact with no valid fields (if that's even possible), it should succeed - and the user will be blocked from sending later with an explicit error message that at least one phone number, email or address should be defined.

**Comment by charlesmchen-signal** [comment]
This PR is now ready for review. 

PTAL @michaelkirk

File: SignalMessaging/attachments/ApproveContactShareViewController.swift [hunk 5]
```diff
@@ -174,13 +183,18 @@ class ContactShareFieldView: UIView {
         }
         field.setIsIncluded(!field.isIncluded())
         checkbox.isSelected = field.isIncluded()
+
+        if let delegate = delegate {
+            delegate.contactShareFieldViewDidChangeSelectedState()
```
Comment (michaelkirk-signal) [inline SignalMessaging/attachments/ApproveContactShareViewController.swift]: more idiomatic: `delegate?. contactShareFieldViewDidChangeSelectedState()`

Comment (charlesmchen-signal) [reply SignalMessaging/attachments/ApproveContactShareViewController.swift]: Done.

**Pull request merged**